spec/fastsync/version2/scheduler.tla

```diff
@@ -0,0 +1,532 @@
+  ELSE
+    LET maxH == maxHeightScheduler(sc) IN
+    maxH > 0 /\ (sc.height >= maxH)
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The behavior of the scheduler state machine                                                   *)
+(* See scheduler.go                                                                              *)
+(* https://github.com/tendermint/tendermint/blockchain/v2/scheduler.go                           *)
+(* ----------------------------------------------------------------------------------------------*)
+InitSc ==
+  /\ scRunning = TRUE
+  /\ outEvent = noEvent
+  /\ \E startHeight \in Heights:
+    scheduler = [
+      initHeight |-> startHeight,
+      height |-> startHeight,
+      peers |-> {},
+      peerHeights |-> [p \in PeerIDs |-> None],
+      peerStates |-> [p \in PeerIDs |-> "peerStateUnknown"],
+      blocks |-> {},
+      blockStates |-> [h \in Heights |-> "blockStateUnknown"],
+      pendingBlocks |-> [h \in Heights |-> None],
+      receivedBlocks |-> [h \in Heights |-> None]
+      ]
+     
+handleAddNewPeer ==
+  /\ inEvent.type = "bcAddNewPeer"
+  /\ LET res == addPeer(scheduler, inEvent.peerID) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+       /\ UNCHANGED <<scheduler>>
+     ELSE
+       /\ scheduler' = res.val
+       /\ UNCHANGED outEvent
+  /\ UNCHANGED scRunning
+       
+handleRemovePeer ==
+  /\ inEvent.type = "bcRemovePeer"
+  /\ LET res == removePeer(scheduler, inEvent.peerID) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scSchedulerFail", reason |-> res.err]
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ scheduler' = res.val
+       /\ IF allBlocksProcessed(scheduler') THEN
+            outEvent' = [type |-> "scFinishedEv", reason |-> "errAfterPeerRemove"]
+          ELSE 
+            UNCHANGED outEvent
+  /\ UNCHANGED scRunning
+           
+handleStatusResponse ==
+  /\ inEvent.type = "bcStatusResponse"
+  /\ LET res == setPeerHeight(scheduler, inEvent.peerID, inEvent.height) IN
+     IF res.err /= noErr THEN
+       /\ outEvent' = [type |-> "scPeerError", peerID |-> inEvent.peerID, reason |-> res.err]
+       /\ UNCHANGED scheduler
+     ELSE
+       /\ scheduler' = res.val
+       /\ outEvent' = noEvent
+  /\ UNCHANGED scRunning
+
+selectPeer(sc, minH) ==
+  LET highPeers == {p \in sc.peers: sc.peerHeights[p] >= minH} IN
+  IF Cardinality(highPeers) = 0 THEN None ELSE CHOOSE p \in highPeers: TRUE
+
+handleTrySchedule == \* every 10 ms, but our spec is asynchronous
+  /\ inEvent.type = "rTrySchedule"
+  /\ LET minH == nextHeightToSchedule(scheduler) IN
+     IF minH = None THEN
+       /\ outEvent'.type = noEvent
+       /\ UNCHANGED scheduler
+     ELSE
+       IF minH = ultimateHeight+1 THEN
+         /\ outEvent' = noEvent
+         /\ UNCHANGED scheduler
+       ELSE
+         /\ LET bestPeerID == selectPeer(scheduler, minH) IN
```

Comment:
good to know. I changed it now but it will need some update as it diverges from the code at this point (there we have a more complex logic of picking peers)